### ISSUE TITLE
struct update syntax causes ICE

### ISSUE THREAD

**Issue by gautamg795** [rust-lang/rust#91502]
### Code

```Rust
use once_cell::sync::OnceCell;
struct Foo {
    a: u64,
    b: u64,
    c: u64,
}
struct Bar {
    b: u64,
    c: u64,

}

static BAR: OnceCell<Bar> = OnceCell::new();

impl Foo {
    fn with_a(a: u64) -> Foo {
        Foo {
            a,
            ..BAR.get().unwrap() // Realized later that this shouldn't compile, but I'd expect an error, not an ICE
        }
    }
}
```
[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=730f6b96dc286b2cb6dcad243fe6af7e)

### Meta

`rustc --version --verbose`:
```
rustc 1.59.0-nightly (acbe4443c 2021-12-02) running on x86_64-unknown-linux-gnu
```

### Error output

```
error: internal compiler error: compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs:321:26: while adjusting Expr { hir_id: HirId { owner: DefId(0:15 ~ playground[29ea]::{impl#0}::with_a), local_id: 6 }, kind: Path(Resolved(None, Path { span: src/main.rs:19:15: 19:18 (#0), res: Def(Static, DefId(0:13 ~ playground[29ea]::BAR)), segments: [PathSegment { ident: BAR#0, hir_id: Some(HirId { owner: DefId(0:15 ~ playground[29ea]::{impl#0}::with_a), local_id: 5 }), res: Some(Err), args: None, infer_args: true }] })), span: src/main.rs:19:15: 19:18 (#0) }, can't compose [Borrow(Ref('_#0r, Not)) -> &once_cell::sync::OnceCell<Bar>] and [Borrow(Ref('_#3r, Not)) -> &once_cell::sync::OnceCell<Bar>]

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1170:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.59.0-nightly (acbe4443c 2021-12-02) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `<impl at src/main.rs:15:1: 22:2>::with_a`
#1 [typeck_item_bodies] type-checking all item bodies
end of query stack
error: could not compile `playground`
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1170:9
stack backtrace:
   0: std::panicking::begin_panic::<rustc_errors::ExplicitBug>
   1: std::panic::panic_any::<rustc_errors::ExplicitBug>
   2: <rustc_errors::HandlerInner>::bug
   3: <rustc_errors::Handler>::bug
   4: rustc_middle::ty::context::tls::with_opt::<rustc_middle::util::bug::opt_span_bug_fmt<rustc_span::span_encoding::Span>::{closure#0}, ()>
   5: rustc_middle::util::bug::opt_span_bug_fmt::<rustc_span::span_encoding::Span>
   6: rustc_middle::util::bug::bug_fmt
   7: <rustc_typeck::check::fn_ctxt::FnCtxt>::apply_adjustments
   8: <rustc_typeck::check::method::confirm::ConfirmContext>::confirm
   9: <rustc_typeck::check::fn_ctxt::FnCtxt>::lookup_method
  10: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
  11: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  12: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
  13: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  14: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
  15: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  16: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_block_with_expected
  17: <rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  18: rustc_typeck::check::check::check_fn
  19: <rustc_infer::infer::InferCtxtBuilder>::enter::<&rustc_middle::ty::context::TypeckResults, <rustc_typeck::check::inherited::InheritedBuilder>::enter<rustc_typeck::check::typeck_with_fallback<rustc_typeck::check::typeck::{closure#0}>::{closure#1}, &rustc_middle::ty::context::TypeckResults>::{closure#0}>
  20: rustc_typeck::check::typeck
  21: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::LocalDefId, &rustc_middle::ty::context::TypeckResults>>
  22: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck
  23: <rustc_middle::hir::map::Map>::par_body_owners::<rustc_typeck::check::typeck_item_bodies::{closure#0}>
  24: rustc_typeck::check::typeck_item_bodies
  25: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), ()>>
  26: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::typeck_item_bodies, rustc_query_impl::plumbing::QueryCtxt>
  27: <rustc_session::session::Session>::time::<(), rustc_typeck::check_crate::{closure#7}>
  28: rustc_typeck::check_crate
  29: rustc_interface::passes::analysis
  30: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorReported>>>
  31: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::analysis, rustc_query_impl::plumbing::QueryCtxt>
  32: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorReported>>
  33: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorReported>>
  34: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
  35: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorReported>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_errors::ErrorReported>>
```

</p>
</details>

**Comment by gautamg795** [comment]
It's likely possible to minimize this further; the snippet posted is just a sanitized version of the actual code that caused the issue. 

Changing the `with_a` function to first assign the `OnceCell` to a local variable does not result in a compiler error and prints the expected error message about a type mismatch:
```rust
    fn with_a(a: u64) -> Foo {
        let bar = BAR.get().unwrap();
        Foo {
            a,
            ..bar
        }
    }
```

**Comment by gautamg795** [comment]
Slightly further minimized:
```rust
use once_cell::sync::OnceCell;
struct Bar {}

static BAR: OnceCell<Bar> = OnceCell::new();

fn main() {
    let b = Bar {
        ..BAR.get().unwrap()
    };
}
```

**Comment by compiler-errors** [comment]
I'd like to take a look at this, I'll take a stab at it tonight.

@rustbot claim

**Comment by compiler-errors** [comment]
I git-blamed this down to 926892ddc0b75 (#90035). We have a "type changing struct-update" here because we're doing `..struct` on a `&Ty`.

I can demote this `bug!()` into a `delay_span_bug` to suppress the ICE, but I'm gonna take a better look at this first and see if I can fix it a bit more cleanly.

**Comment by Badel2** [comment]
I run into this while fuzzing, so if you need another example:

```rust
fn main() {
    Ok { ..0 > 0 }
}
```

**Comment by Alexendoo** [comment]
Fixed by #92112
